src/s3path.jl

```diff
@@ -68,25 +73,27 @@ function S3Path(
     key::AbstractPath;
     isdirectory::Bool=false,
     config::AbstractAWSConfig=global_aws_config(),
+    version::S3PathVersion=nothing,
 )
     return S3Path(
         key.segments,
         "/",
         normalize_bucket_name(bucket),
         isdirectory,
         config,
+        version,
     )
 end
 
 # To avoid a breaking change.
-function S3Path(str::AbstractString; config::AbstractAWSConfig=global_aws_config())
-    result = tryparse(S3Path, str; config=config)
+function S3Path(str::AbstractString; config::AbstractAWSConfig=global_aws_config(), version=nothing)
+    result = tryparse(S3Path, str; config=config, version=version)
     result !== nothing || throw(ArgumentError("Invalid s3 path string: $str"))
     return result
 end
 
 # if config=nothing, will not try to talk to AWS until after string is confirmed to be an s3 path
-function Base.tryparse(::Type{<:S3Path}, str::AbstractString; config::Union{Nothing,AbstractAWSConfig}=nothing)
+function Base.tryparse(::Type{<:S3Path}, str::AbstractString; config::Union{Nothing,AbstractAWSConfig}=nothing, version::S3PathVersion=nothing)
```

Comment:
I feel like in practise having the version as part of the parsed string would be the most useful. We could probably use the HTTP GET parameter format which is already used by S3 HTTP requests: https://docs.aws.amazon.com/AmazonS3/latest/userguide/RetMetaOfObjVersion.html

Probably the easiest approach to this is to leave out `version` support here unless you need this right away and we can figure out the right approach independent from the rest of the support.

Reply:
Oh interesting. I agree that we should additionally support having the version as part of the parsed string, but I think having this current implementation is orthogonally important. 

E.g., without it, we're back to needing to construct all of our paths like so:
```
my_path = S3Path("my_bucket", "my_key"; version="my_version")
```
instead of
```
my_path = S3Path("s3://my_bucket/my_key"; version="my_version")
```
which makes it a (slight) pain to pass around variables-per-object.  

What would you think about keeping the current implementation here and then (in a separate PR, I can open an issue) additionally supporting parsing the version _from_ `str`. (Probably something like 
```
str = "s3://my_bucket/my_key?versionId=L4kqtJlcpXroDTDmpUMLUo"
```
, a la the REST API command: https://docs.aws.amazon.com/AmazonS3/latest/userguide/RetrievingObjectVersions.html)

...I also think this would be straightforward enough to add/add tests for here.

Reply:
I'll call out I don't think
```julia
S3Path("s3://my_bucket/my_key"; version="my_version")
```
currently works since the above `S3Path` constructor doesn't take a `version` keyword: https://github.com/JuliaCloud/AWSS3.jl/pull/161/files#diff-31b593db3c82f33730f4af58846ed59213199a8dd973bddc4f70f1e1721d6719R89

Maybe a good path forward is to:
- Add the `version` keyword to `S3Path(::AbstractString; config::AbstractAWSConfig)`
- Remove the `version` keyword from `tryparse`. This avoids having to deprecate the keyword here in the future

To do this we'd need to either make an `S3Path` without a version and then make a new one with the version added in or revise `tryparse` to call a private function that returns the parsed components

Reply:
> I'll call out I don't think 
> [...]
> currently works since the above S3Path constructor doesn't take a version keyword:

...ah right, b/c we literally just removed it. :| 

> Maybe a good path forward is to:
> - Add the version keyword to S3Path(::AbstractString; config::AbstractAWSConfig)
> - Remove the version keyword from tryparse. This avoids having to deprecate the keyword in the future for tryparse
To do this we'd need to either make an S3Path without a version and then make a new one with the version added in or revise tryparse to call a private function that returns the parsed components

Yeah, this sounds good to me, I'll make that change. I think prefer your first approach, so I'll go with that.

Reply:
Righto, change made---see what you think. For future safety, I threw in an error in the case that both a version and a parseable-string-containing-a-version are passed in, although we won't ever hit that error until we do the follow-up work of parsing the version from the string in the first place.